File: musicApp/Sync/QueueSync.swift
import Foundation
import Combine
import FirebaseFirestore

/// Queue reconciliation.
///
/// Two write shapes:
///  - **Intent ops** (`insert`/`remove`/`move`/`consumeHead`): rebased inside a
///    transaction against the *current* remote queue, anchored by trackID — never
///    by index. Two devices editing concurrently both land; a stale index can't
///    delete the wrong row.
///  - **`replaceAll`**: LWW bulk write for observed local edits (drag-reorder,
///    clear). Live edits win unconditionally (they are the newest user intent);
///    *offline replays* carry the version basis they were built on and lose to
///    anything newer (`SyncError.queueStale`) — an offline device must not
///    resurrect a queue the rest of the session already moved past.
final class QueueSync {

    private let db: Firestore
    private let sessionRef: () -> DocumentReference?

    /// Single-slot offline outbox: (queue, version basis it was derived from).
    private var pendingReplace: (queue: [TrackRef], basis: Int)?
    private var bag = Set<AnyCancellable>()

    init(db: Firestore, sessionRef: @escaping () -> DocumentReference?,
         isOnline: AnyPublisher<Bool, Never>) {
        self.db = db
        self.sessionRef = sessionRef
        isOnline
            .removeDuplicates()
            .filter { $0 }
            .sink { [weak self] _ in self?.flushPending() }
            .store(in: &bag)
    }

    // MARK: - Public API

    /// `isReplay` is passed explicitly by flushPending — inferring it from
    /// pendingReplace was a bug: a LIVE edit made while an offline replay sat
    /// buffered would inherit replay semantics and could be discarded as stale.
    func apply(_ op: QueueOp, basisVersion: Int, isReplay: Bool = false) async {
        guard let ref = sessionRef() else { return }
        let dev = SyncDevice.id
        do {
            try await db.txn { txn in
                let snap = try txn.getDocument(ref)
                guard let cur = SessionState(snap: snap) else { throw SyncError.corrupt }

                // Stale-replay guard applies only to bulk overwrites.
                if case .replaceAll = op, isReplay, cur.queueVersion != basisVersion {
                    throw SyncError.queueStale
                }
                guard let rebased = Self.rebase(op, onto: cur.queue) else { return } // no-op
                txn.updateData([
                    "queue": rebased.map(\.dict),
                    "queueVersion": cur.queueVersion + 1,
                    "updatedBy": dev,
                ], forDocument: ref)
            }
            pendingReplace = nil
        } catch SyncError.queueStale {
            pendingReplace = nil               // we lost — remote queue is newer intent
            print("🗑 [QueueSync] Offline queue replay discarded (stale basis)")
        } catch is SyncError {
            pendingReplace = nil
        } catch {
            // Offline / transient. Only bulk state is worth buffering; a lost
            // single op offline is recoverable by the user, a wrong bulk
            // overwrite on reconnect is not.
            if case .replaceAll(let q) = op {
                pendingReplace = (q, basisVersion)
            }
        }
    }

    private func flushPending() {
        guard let (q, basis) = pendingReplace else { return }
        Task { await apply(.replaceAll(q), basisVersion: basis, isReplay: true) }
    }

    // MARK: - Rebase (pure — unit-testable without Firestore)

    /// Returns nil when the op is a no-op against the current queue
    /// (target vanished, CAS failed). Anchors resolve by trackID.
    static func rebase(_ op: QueueOp, onto queue: [TrackRef]) -> [TrackRef]? {
        switch op {
        case .insert(let ref, let afterID):
            var q = queue
            q.insert(ref, at: insertionIndex(afterID: afterID, in: q))
            return q

        case .remove(let id):
            guard let idx = queue.firstIndex(where: { $0.id == id }) else { return nil }
            var q = queue
            q.remove(at: idx)
            return q

        case .move(let id, let afterID):
            guard let idx = queue.firstIndex(where: { $0.id == id }) else { return nil }
            var q = queue
            let item = q.remove(at: idx)
            q.insert(item, at: insertionIndex(afterID: afterID, in: q))
            return q

        case .consumeHead(let expected):
            // CAS: only pop if the head is still what the owner played.
            // If a follower reordered/inserted meanwhile, their edit survives.
            guard queue.first?.id == expected else { return nil }
            return Array(queue.dropFirst())

        case .replaceAll(let q):
            return q
        }
    }

    private static func insertionIndex(afterID: UUID?, in queue: [TrackRef]) -> Int {
        guard let afterID else { return 0 }                                   // nil = front
        guard let idx = queue.firstIndex(where: { $0.id == afterID }) else {
            return queue.count                                                // anchor gone → append
        }
        return idx + 1
    }
}

File: desktop/src/queueSync.ts
// Twin of Sync/QueueSync.swift — intent ops rebased by trackID, CAS consumeHead,
// version-basis guard for offline replaceAll replays.
import { Firestore, DocumentReference, runTransaction } from "firebase/firestore";
import { QueueOp, TrackRef, SessionState, DEVICE_ID, QUEUE_STALE, sameId } from "./protocol";

export class QueueSync {
  private pendingReplace?: { queue: TrackRef[]; basis: number };

  constructor(
    private db: Firestore,
    private ref: () => DocumentReference | undefined,
  ) {}

  onOnline() { this.flushPending(); }

  /** `isReplay` comes only from flushPending — inferring it from pendingReplace
   *  made a LIVE edit inherit replay semantics while an offline replay sat
   *  buffered, so it could be discarded as stale. */
  async apply(op: QueueOp, basisVersion: number, isReplay = false) {
    const ref = this.ref();
    if (!ref) return;
    try {
      await runTransaction(this.db, async txn => {
        const snap = await txn.get(ref);
        const cur = snap.data() as SessionState | undefined;
        if (!cur) throw new Error("corrupt");
        if (op.kind === "replaceAll" && isReplay && cur.queueVersion !== basisVersion)
          throw QUEUE_STALE;
        const rebased = rebase(op, cur.queue);
        if (!rebased) return;                                  // no-op
        txn.update(ref, {
          queue: rebased, queueVersion: cur.queueVersion + 1, updatedBy: DEVICE_ID,
        });
      });
      this.pendingReplace = undefined;
    } catch (e) {
      if (e === QUEUE_STALE) {
        this.pendingReplace = undefined;                       // remote is newer intent
        console.log("[queueSync] offline replay discarded (stale basis)");
      } else if (op.kind === "replaceAll") {
        this.pendingReplace = { queue: op.queue, basis: basisVersion };
      }
      // lost single ops offline are user-recoverable; wrong bulk replays are not
    }
  }

  private flushPending() {
    const p = this.pendingReplace;
    if (p) void this.apply({ kind: "replaceAll", queue: p.queue }, p.basis, true);
  }
}

/** Pure — unit-testable without Firestore. null = no-op. */
export function rebase(op: QueueOp, queue: TrackRef[]): TrackRef[] | null {
  const idx = (id: string) => queue.findIndex(t => sameId(t.id, id));
  const insertAt = (afterId: string | null, q: TrackRef[]) => {
    if (afterId === null) return 0;
    const i = q.findIndex(t => sameId(t.id, afterId));
    return i < 0 ? q.length : i + 1;                           // anchor gone → append
  };
  switch (op.kind) {
    case "insert": {
      const q = [...queue];
      q.splice(insertAt(op.afterId, q), 0, op.ref);
      return q;
    }
    case "remove": {
      const i = idx(op.id);
      if (i < 0) return null;
      const q = [...queue]; q.splice(i, 1); return q;
    }
    case "move": {
      const i = idx(op.id);
      if (i < 0) return null;
      const q = [...queue];
      const [item] = q.splice(i, 1);
      q.splice(insertAt(op.afterId, q), 0, item);
      return q;
    }
    case "consumeHead":
      // CAS: only pop if the head is what the owner actually played.
      return queue[0] && sameId(queue[0].id, op.expected) ? queue.slice(1) : null;
    case "consumeHeadRun": {
      // All-or-nothing CAS over the whole run: a follower that inserted
      // anywhere inside it wins, exactly like single consumeHead.
      const ids = op.expected;
      if (!ids.length || queue.length < ids.length) return null;
      if (!ids.every((id, i) => sameId(queue[i].id, id))) return null;
      return queue.slice(ids.length);
    }
    case "replaceAll":
      return op.queue;
    case "append":
      // Bulk enqueue — twin of queuePlaylist's queue.append(contentsOf:).
      return op.refs.length ? [...queue, ...op.refs] : null;
    case "removeMany": {
      // Rebased per id: ones already gone elsewhere simply don't match, so a
      // concurrent removal of the same track isn't an error.
      const drop = new Set(op.ids.map(id => id.toLowerCase()));
      const q = queue.filter(t => !drop.has(t.id.toLowerCase()));
      return q.length === queue.length ? null : q;
    }
    case "injectFront": {
      // Twin of injectAtFrontOfQueue: pull the set out of wherever it sits,
      // then plant the not-now-playing remainder at the head.
      const rm = new Set(op.removeIds.map(id => id.toLowerCase()));
      return [...op.refs, ...queue.filter(t => !rm.has(t.id.toLowerCase()))];
    }
  }
}
